### ISSUE TITLE
Question: How Can I Implement Call Graph Analysis in a Spring Boot Project?

### ISSUE THREAD

**Issue by wangxi761** [soot-oss/SootUp#639]
I am looking to analyze the call graph for my project, but I've encountered some issues. Could you please assist me in resolving them?

1. For call graph analysis, I need to define an entry method. However, in Java web projects, such as those using the Spring Boot framework, if I use the main method for analysis, it only contains methods related to the main function. The method invocation within the Spring framework isn't as intuitive as I would like.
2. Is it possible to analyze all methods without specifying an entry method?
3. How can I analyze the call graph for my project code while filtering out all other library code? Is it possible for me to customize this?
4. How should I go about analyzing dependency injection within the Spring framework?

Thank you in advance for your help.

**Comment by JonasKlauke** [comment]
Hi call graph creation of spring boot applications is a open (research) problem, because spring connects the methods using a lot of dynamic java features, which are hard to resolve statically. Calls are also created by annotations or configuration files which makes it really hard to resolve.

Another problem you might face with sootUp is that spring boot uses its own class loader and the internal structure of the fatjar is a little different. This will cause that the package name of the classes will be different. Your application code will be in the folder BOOT-INF/classes. If you don't modify the jar before the analysis their fully qualified name will be BOOT-INF.classes.com.example.A

To analyze all methods or all public methods, you can simply iterate through all classes and create a long entry methods list. You can use the package name to find out which class is your class.

**Comment by wangxi761** [comment]
for the fat jar, maybe i can contribute a  ` FatJarAnalysisInputLocation ` to handle this situation. 

But in engineering practice, it's unlikely that we'll be able to customize entry point it for every project ，and i need to process different project batch . so  i want it to be possible to analyze all method calls and produce a call graph even when there is no entry point .  After that, I can use uniform rules for processing the call graph.   What do you think about it？

**Comment by StevenArzt** [comment]
We have an implementation of a Spring Boot analyzer in our commercial scanner based on Soot (the old one, not SootUp). The toughest challenge was to understand all the ways in which Sprint Boot can be configured: via annotation, via XML files, via API calls. We then had to parse of these fragments, create a consistent model of the application, and resolve the dependencies that would normally be injected by Spring Boot. This is an iterative process because dependencies can be nested. On the plus side, we also get the external entry points (e.g., REST endpoints), which we put in FlowDroid's dummy main method.

The bottom line is essentially that it's doable, but a ton of work.

**Comment by JonasKlauke** [comment]
It would be nice if you could develop a spring boot specific AnalysisInputLocation, since spring boot creates fatjars that contain nested Jars of dependencies and the application classes are moved to a different location. It would be nice to simply pass these fatjars and the view will contain everything.

To come back to your method problem.
You can create specific views which only use a specific application jar. On the view you can call getClasses() to get every class in the jar. For each of the classes you can call getmethods() to get all methods of the class

**Comment by JonasKlauke** [comment]
Maybe you misunderstood entry method.
For the call graph algorithm these are simple start points to put in the Worklist to process them

**Comment by RespectandLove** [comment]
First  thank you for your answer. However, the reason for the question is that I used Spring Boot. After reviewing the source code of SoutUp, I found that it cannot generate Call Graph
